File: nqp-truffle/src/main/java/org/raku/nqp/truffle/RxWire.java
package org.raku.nqp.truffle;

/**
 * The wire format a compiled regex descriptor travels in.
 *
 * <p>The backend flattens a QAST::Regex tree to an int array plus a pool of
 * strings (see {@code src/vm/jvm/QAST/RxDescriptor.nqp}), and the class file
 * has to carry that across to the engine. A string constant is the cheapest
 * thing bytecode can hold: one {@code ldc} and no array-building code at
 * every regex, and it doubles as the Truffle {@code Source} the language
 * parses, so a descriptor gets its own call target and its own compiled code
 * exactly as a pattern does.
 *
 * <p>The format is
 *
 * <pre>{@code
 *   rxd <scan> <codeLen> <int>... <poolLen> (<len>:<chars>)... <len>:<passName>
 * }</pre>
 *
 * <p>Pool entries are length-prefixed rather than delimited because a regex
 * may match any character at all, including whichever one a delimiter would
 * have been.
 *
 * <p>The trailing entry is the name {@code !cursor_pass} is called with,
 * which is what makes it reduce and build the match tree; empty means the
 * rule does not reduce.
 */
public final class RxWire {

    /** Marks a source as a descriptor rather than a pattern to parse. */
    public static final String MAGIC = "rxd ";

    private RxWire() { }

    /**
     * @param passName what {@code !cursor_pass} is called with, empty when
     *                 the rule does not reduce.
     * @param scan     whether the rule may retry at later start positions.
     *                 It is a property of the rule rather than an opcode
     *                 because the retry loop has to update {@code $!from} on
     *                 the cursor, which a program cannot do.
     */
    public record Descriptor(int[] code, Object[] pool, String passName, boolean scan) { }

    public static boolean isDescriptor(String source) {
        return source.startsWith(MAGIC);
    }

    public static String encode(int[] code, Object[] pool, String passName, boolean scan) {
        StringBuilder out = new StringBuilder(MAGIC);
        out.append(scan ? 1 : 0).append(' ');
        out.append(code.length);
        for (int c : code) out.append(' ').append(c);
        out.append(' ').append(pool.length);
        for (Object entry : pool) chunk(out, String.valueOf(entry));
        chunk(out, passName);
        return out.toString();
    }

    private static void chunk(StringBuilder out, String entry) {
        out.append(' ').append(entry.length()).append(':').append(entry);
    }

    public static Descriptor decode(String source) {
        if (!isDescriptor(source)) {
            throw new IllegalArgumentException("not a regex descriptor");
        }
        Reader in = new Reader(source, MAGIC.length());

        boolean scan = in.nextInt() != 0;
        int[] code = new int[in.nextInt()];
        for (int i = 0; i < code.length; i++) code[i] = in.nextInt();

        Object[] pool = new Object[in.nextInt()];
        for (int i = 0; i < pool.length; i++) pool[i] = in.nextChunk();

        return new Descriptor(code, pool, in.nextChunk(), scan);
    }

    /** Reads the format's two token shapes, and insists the input is well formed. */
    private static final class Reader {
        private final String src;
        private int at;

        Reader(String src, int at) { this.src = src; this.at = at; }

        int nextInt() {
            int start = at;
            if (at < src.length() && src.charAt(at) == '-') at++;
            while (at < src.length() && isDigit(src.charAt(at))) at++;
            if (at == start) throw bad("expected an integer");
            int value = Integer.parseInt(src, start, at, 10);
            skipSpace();
            return value;
        }

        /** A {@code <len>:<chars>} pool entry. */
        String nextChunk() {
            int start = at;
            while (at < src.length() && isDigit(src.charAt(at))) at++;
            if (at == start) throw bad("expected a pool entry length");
            int len = Integer.parseInt(src, start, at, 10);
            if (at >= src.length() || src.charAt(at) != ':') throw bad("expected ':'");
            at++;
            if (at + len > src.length()) throw bad("pool entry runs past the end");
            String value = src.substring(at, at + len);
            at += len;
            skipSpace();
            return value;
        }

        private void skipSpace() {
            if (at < src.length() && src.charAt(at) == ' ') at++;
        }

        private static boolean isDigit(char c) { return c >= '0' && c <= '9'; }

        private IllegalArgumentException bad(String what) {
            return new IllegalArgumentException(
                "malformed regex descriptor at " + at + ": " + what);
        }
    }
}

File: nqp-truffle/src/main/kotlin/org/raku/nqp/truffle/RxWire.kt
package org.raku.nqp.truffle

/**
 * The wire format a compiled regex descriptor travels in.
 *
 * The backend flattens a QAST::Regex tree to an int array plus a pool of
 * strings (see `src/vm/jvm/QAST/RxDescriptor.nqp`), and the class file has to
 * carry that across to the engine. A string constant is the cheapest thing
 * bytecode can hold: one `ldc` and no array-building code at every regex, and
 * it doubles as the Truffle `Source` the language parses, so a descriptor
 * gets its own call target and its own compiled code exactly as a pattern
 * does.
 *
 * The format is
 *
 * ```
 *   rxd <scan> <codeLen> <int>... <poolLen> (<len>:<chars>)... <len>:<passName>
 * ```
 *
 * Pool entries are length-prefixed rather than delimited because a regex may
 * match any character at all, including whichever one a delimiter would have
 * been.
 *
 * The trailing entry is the name `!cursor_pass` is called with, which is what
 * makes it reduce and build the match tree; empty means the rule does not
 * reduce.
 */
object RxWire {

    /** Marks a source as a descriptor rather than a pattern to parse. */
    const val MAGIC = "rxd "

    /**
     * @param passName what `!cursor_pass` is called with, empty when the rule
     *   does not reduce.
     * @param scan whether the rule may retry at later start positions. It is
     *   a property of the rule rather than an opcode because the retry loop
     *   has to update `$!from` on the cursor, which a program cannot do.
     * @param resumable whether the rule passes with :backtrack, so that
     *   `!cursor_next` may re-enter it for its next match; the engine keeps
     *   its choice points after a pass to answer that.
     */
    @JvmRecord
    data class Descriptor(
        val code: IntArray,
        val pool: Array<Any?>,
        val passName: String,
        val scan: Boolean,
        val resumable: Boolean,
    )

    @JvmStatic
    fun isDescriptor(source: String): Boolean = source.startsWith(MAGIC)

    @JvmStatic
    @JvmOverloads
    fun encode(
        code: IntArray, pool: Array<Any?>, passName: String, scan: Boolean,
        resumable: Boolean = false,
    ): String {
        val out = StringBuilder(MAGIC)
        out.append((if (scan) 1 else 0) or (if (resumable) 2 else 0)).append(' ')
        out.append(code.size)
        for (c in code) out.append(' ').append(c)
        out.append(' ').append(pool.size)
        for (entry in pool) chunk(out, entry.toString())
        chunk(out, passName)
        return out.toString()
    }

    private fun chunk(out: StringBuilder, entry: String) {
        out.append(' ').append(entry.length).append(':').append(entry)
    }

    @JvmStatic
    fun decode(source: String): Descriptor {
        require(isDescriptor(source)) { "not a regex descriptor" }
        val input = Reader(source, MAGIC.length)

        val flags = input.nextInt()
        val code = IntArray(input.nextInt())
        for (i in code.indices) code[i] = input.nextInt()

        val pool = arrayOfNulls<Any?>(input.nextInt())
        for (i in pool.indices) pool[i] = input.nextChunk()

        return Descriptor(
            code, pool, input.nextChunk(),
            (flags and 1) != 0, (flags and 2) != 0,
        )
    }

    /** Reads the format's two token shapes, and insists the input is well formed. */
    private class Reader(private val src: String, private var at: Int) {

        fun nextInt(): Int {
            val start = at
            if (at < src.length && src[at] == '-') at++
            while (at < src.length && isDigit(src[at])) at++
            if (at == start) throw bad("expected an integer")
            val value = src.substring(start, at).toInt()
            skipSpace()
            return value
        }

        /** A `<len>:<chars>` pool entry. NFG: len is a grapheme count (the
         * encoder's nqp::chars), so consume that many graphemes. Grapheme count
         * is invariant under NFC, so segmenting the raw descriptor matches. */
        fun nextChunk(): String {
            val start = at
            while (at < src.length && isDigit(src[at])) at++
            if (at == start) throw bad("expected a pool entry length")
            val len = src.substring(start, at).toInt()
            if (at >= src.length || src[at] != ':') throw bad("expected ':'")
            at++
            val end = graphemeEnd(src, at, len)
            if (end > src.length) throw bad("pool entry runs past the end")
            val value = src.substring(at, end)
            at = end
            skipSpace()
            return value
        }

        /** The UTF-16 offset [count] graphemes past [from] in [s]. */
        private fun graphemeEnd(s: String, from: Int, count: Int): Int {
            if (count <= 0) return from
            val bi = java.text.BreakIterator.getCharacterInstance()
            bi.setText(s)
            var pos = from
            var n = count
            while (n > 0) {
                val next = bi.following(pos)
                if (next == java.text.BreakIterator.DONE) return s.length
                pos = next
                n--
            }
            return pos
        }

        private fun skipSpace() {
            if (at < src.length && src[at] == ' ') at++
        }

        private fun isDigit(c: Char): Boolean = c in '0'..'9'

        private fun bad(what: String) =
            IllegalArgumentException("malformed regex descriptor at $at: $what")
    }
}
